src/v/storage/disk_log_impl.cc

```diff
@@ -1972,6 +1972,203 @@ ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
     };
 }
 
+ss::future<log::offset_range_size_result_t> disk_log_impl::offset_range_size(
+  model::offset first,
+  offset_range_size_requirements_t target,
+  ss::io_priority_class io_priority) {
+    if (
+      false
+      && target.target_size
+           < storage::segment_index::default_data_buffer_step) {
+        vlog(
+          stlog.error,
+          "The target size {}/{} is below segment index default data buffer "
+          "step ({})",
+          target.target_size,
+          target.min_size,
+          storage::segment_index::default_data_buffer_step);
+        throw std::invalid_argument("Target size is too small");
+    }
+
+    std::vector<ss::lw_shared_ptr<storage::segment>> segments;
+    // Last offset included to the result, default value means that we didn't
+    // find anything
+    model::offset last_included_offset;
+    auto base_it = _segs.lower_bound(first);
+
+    // Invariant: 'first' offset should be present in the log. If the segment is
+    // compacted it's OK if it's missing if the segment that stores the offset
+    // range that includes the offset still exists.
+    if (base_it == _segs.end()) {
+        vlog(
+          stlog.error,
+          "Offset {} is not present in the log {}",
+          first,
+          offsets());
+        throw std::invalid_argument("Offset out of range");
+    } else {
+        auto lstat = base_it->get()->offsets();
+        if (first < lstat.base_offset || first > lstat.committed_offset) {
+            vlog(
+              stlog.error,
+              "Offset {} is not present in the segment {}",
+              first,
+              lstat);
+            throw std::invalid_argument("Offset out of range");
+        }
+    }
+
+    // Collect relevant segments
+    size_t current_size = 0;
+    size_t base_file_pos = 0;
+    for (auto it = base_it; it < _segs.end(); it++) {
+        const auto& offsets = it->get()->offsets();
+        auto r_lock = co_await it->get()->read_lock();
+        if (offsets.base_offset <= first && offsets.committed_offset >= first) {
```

Comment:
generally speaking anytime we have a reference to something accessed via an iterator held across `co_await` calls we should see a comment that explains why it is safe. can you add that here?